File: app/src/main/java/com/kickstarter/viewmodels/SurveyResponseViewModel.java
package com.kickstarter.viewmodels;

import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.models.SurveyResponse;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.SurveyResponseActivity;

import androidx.annotation.NonNull;
import okhttp3.Request;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.ignoreValues;

public interface SurveyResponseViewModel {

  interface Inputs {
    /** Call when the dialog's OK button has been clicked. */
    void okButtonClicked();

    /** Call when a project uri request has been made. */
    void projectUriRequest(Request request);

    /** Call when a project survey uri request has been made. */
    void projectSurveyUriRequest(Request request);
  }

  interface Outputs {
    /** Emits when we should navigate back. */
    Observable<Void> goBack();

    /** Emits when we should show a confirmation dialog. */
    Observable<Void> showConfirmationDialog();

    /** Emits a url to load in the web view. */
    Observable<String> webViewUrl();
  }

  final class ViewModel extends ActivityViewModel<SurveyResponseActivity> implements Inputs, Outputs {

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      final Observable<SurveyResponse> surveyResponse = intent()
        .map(i -> i.getParcelableExtra(IntentKey.SURVEY_RESPONSE))
        .ofType(SurveyResponse.class);

      final Observable<String> surveyWebUrl = surveyResponse
        .map(r -> r.urls().web().survey());

      surveyWebUrl
        .compose(bindToLifecycle())
        .subscribe(this.webViewUrl);

      final Observable<Pair<Request, String>> projectRequestAndSurveyUrl = Observable.combineLatest(
        this.projectUriRequest,
        surveyWebUrl,
        Pair::create
      );

      projectRequestAndSurveyUrl
        .filter(this::requestTagUrlIsSurveyUrl)
        .compose(ignoreValues())
        .compose(bindToLifecycle())
        .subscribe(this.showConfirmationDialog);

      this.goBack = this.okButtonClicked;
    }

    /**
     * Returns if a project request tag's url is a survey url,
     * which indicates a redirect from a successful submit.
     */
    private boolean requestTagUrlIsSurveyUrl(final @NonNull Pair<Request, String> projectRequestAndSurveyUrl) {
      return ((Request) projectRequestAndSurveyUrl.first.tag()).url().toString()
        .equals(projectRequestAndSurveyUrl.second);
    }

    private final PublishSubject<Void> okButtonClicked = PublishSubject.create();
    private final PublishSubject<Request> projectUriRequest = PublishSubject.create();
    private final PublishSubject<Request> projectSurveyUriRequest = PublishSubject.create();

    private final Observable<Void> goBack;
    private final PublishSubject<Void> showConfirmationDialog = PublishSubject.create();
    private final BehaviorSubject<String> webViewUrl = BehaviorSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void okButtonClicked() {
      this.okButtonClicked.onNext(null);
    }
    @Override public void projectUriRequest(final @NonNull Request request) {
      this.projectUriRequest.onNext(request);
    }
    @Override public void projectSurveyUriRequest(final @NonNull Request request) {
      this.projectSurveyUriRequest.onNext(request);
    }

    @Override public @NonNull Observable<Void> goBack() {
      return this.goBack;
    }
    @Override public @NonNull Observable<Void> showConfirmationDialog() {
      return this.showConfirmationDialog;
    }
    @Override public @NonNull Observable<String> webViewUrl() {
      return this.webViewUrl;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/SurveyResponseViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.util.Pair
import androidx.core.net.toUri
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.UrlUtils
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.path
import com.kickstarter.models.SurveyResponse
import com.kickstarter.ui.IntentKey
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import okhttp3.Request

interface SurveyResponseViewModel {
    interface Inputs {
        /** Call when the dialog's OK button has been clicked.  */
        fun okButtonClicked()

        /** Call when a project uri request has been made.  */
        fun projectUriRequest(request: Request)

        /** Call when a project survey uri request has been made.  */
        fun projectSurveyUriRequest(request: Request)
    }

    interface Outputs {
        /** Emits when we should navigate back.  */
        fun goBack(): Observable<Unit>

        /** Emits when we should show a confirmation dialog.  */
        fun showConfirmationDialog(): Observable<Unit>

        /** Emits a url to load in the web view.  */
        fun webViewUrl(): Observable<String>
    }

    class ViewModel(val environment: Environment, val intent: Intent) :
        androidx.lifecycle.ViewModel(), Inputs, Outputs {
        /**
         * Returns if a project request tag's url is a survey url,
         * which indicates a redirect from a successful submit.
         */
        private fun requestTagUrlIsSurveyUrl(projectRequestAndSurveyUrl: Pair<Request, String>): Boolean {
            val tag = projectRequestAndSurveyUrl.first.tag() as Request?
            return tag == null || tag.url.toString() == projectRequestAndSurveyUrl.second
        }

        private val okButtonClicked = PublishSubject.create<Unit>()
        private val projectUriRequest = PublishSubject.create<Request>()
        private val projectSurveyUriRequest = PublishSubject.create<Request>()
        private val goBack: Observable<Unit>
        private val showConfirmationDialog = PublishSubject.create<Unit>()
        private val webViewUrl = BehaviorSubject.create<String>()

        val inputs: Inputs = this
        val outputs: Outputs = this

        private fun intent() = intent.let { Observable.just(it) }

        private val disposables = CompositeDisposable()

        init {

            val surveyActivityUrl = intent()
                .filter {
                    it.hasExtra(IntentKey.SURVEY_RESPONSE) && it.getParcelableExtra<SurveyResponse>(IntentKey.SURVEY_RESPONSE).isNotNull()
                }
                .map { requireNotNull(it.getParcelableExtra(IntentKey.SURVEY_RESPONSE)) }
                .ofType(SurveyResponse::class.java)
                .map {
                    it.urls()?.web()?.survey() ?: ""
                }

            val pledgeRedemptionUrl = intent()
                .filter {
                    it.hasExtra(IntentKey.NOTIFICATION_PLEDGE_REDEMPTION) && !it.getStringExtra(IntentKey.NOTIFICATION_PLEDGE_REDEMPTION).isNullOrEmpty()
                }
                .map { requireNotNull(it.getStringExtra(IntentKey.NOTIFICATION_PLEDGE_REDEMPTION)) }
                .ofType(String::class.java)
                .map { UrlUtils.appendPath(environment.webEndpoint(), it) }

            val surveyNotificationUrl = intent()
                .filter {
                    it.hasExtra(IntentKey.NOTIFICATION_SURVEY_RESPONSE) && !it.getStringExtra(IntentKey.NOTIFICATION_SURVEY_RESPONSE)
                        .isNullOrEmpty()
                }
                .map { requireNotNull(it.getStringExtra(IntentKey.NOTIFICATION_SURVEY_RESPONSE)) }
                .ofType(String::class.java)
                .map { UrlUtils.appendPath(environment.webEndpoint(), it) }

            val surveyDeeplinkUrl = intent()
                .filter {
                    it.hasExtra(IntentKey.DEEPLINK_SURVEY_RESPONSE) && !it.getStringExtra(IntentKey.DEEPLINK_SURVEY_RESPONSE)
                        .isNullOrEmpty()
                }
                .map { requireNotNull(it.getStringExtra(IntentKey.DEEPLINK_SURVEY_RESPONSE)) }
                .ofType(String::class.java)
                .map { environment.webEndpoint() + it.toUri().path() }

            val surveyUrl = Observable.merge(surveyActivityUrl, surveyNotificationUrl, surveyDeeplinkUrl, pledgeRedemptionUrl)

            surveyUrl
                .subscribe {
                    webViewUrl.onNext(it)
                }
                .addToDisposable(disposables)

            val projectRequestAndSurveyUrl =
                Observable.combineLatest<Request, String?, Pair<Request, String>>(
                    projectUriRequest,
                    surveyUrl
                ) { a: Request?, b: String? -> Pair.create(a, b) }

            projectRequestAndSurveyUrl
                .filter { projectRequestAndSurveyUrl: Pair<Request, String> ->
                    requestTagUrlIsSurveyUrl(
                        projectRequestAndSurveyUrl
                    )
                }
                .compose(Transformers.ignoreValuesV2())
                .subscribe { showConfirmationDialog.onNext(it) }
                .addToDisposable(disposables)

            goBack = okButtonClicked
        }

        override fun okButtonClicked() {
            okButtonClicked.onNext(Unit)
        }

        override fun projectUriRequest(request: Request) {
            projectUriRequest.onNext(request)
        }

        override fun projectSurveyUriRequest(request: Request) {
            projectSurveyUriRequest.onNext(request)
        }

        override fun goBack(): Observable<Unit> = goBack

        override fun showConfirmationDialog(): Observable<Unit> = showConfirmationDialog

        override fun webViewUrl(): Observable<String> = webViewUrl

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }
    }

    class Factory(private val environment: Environment, private val intent: Intent) : ViewModelProvider.Factory {
        override fun <T : androidx.lifecycle.ViewModel> create(modelClass: Class<T>): T {
            return ViewModel(environment, intent) as T
        }
    }
}
